src/raftstore/store/fsm/batch.rs

```diff
@@ -0,0 +1,575 @@
+        self.normals.clear();
+        self.control.take();
+    }
+
+    /// Put back the FSM located at index.
+    ///
+    /// Only messages located after `channel_pos` will trigger further notification.
+    /// This function may fail if messages is put after `channel_pos` before FSM is released.
+    pub fn release(&mut self, index: usize, channel_pos: usize) -> bool {
+        let mut fsm = self.normals.swap_remove(index);
+        let mailbox = fsm.take_mailbox().unwrap();
+        mailbox.release(fsm);
+        if mailbox.len() == channel_pos {
+            true
+        } else {
+            match mailbox.take_fsm() {
+                None => true,
+                Some(mut s) => {
+                    s.set_mailbox(Cow::Owned(mailbox));
+                    let last_index = self.normals.len();
+                    self.normals.push(s);
+                    self.normals.swap(index, last_index);
+                    false
+                }
+            }
+        }
+    }
+
+    /// Remove the normal FSM located at `index`.
+    ///
+    /// This method should only be called when the FSM is stopped.
+    /// If there are still messages in channel, the FSM is untouched and
+    /// the function will return false to let caller to keep polling.
+    pub fn remove(&mut self, index: usize) -> bool {
+        let mut fsm = self.normals.swap_remove(index);
+        let mailbox = fsm.take_mailbox().unwrap();
+        if mailbox.is_empty() {
+            mailbox.release(fsm);
+            true
+        } else {
+            fsm.set_mailbox(Cow::Owned(mailbox));
+            let last_index = self.normals.len();
+            self.normals.push(fsm);
+            self.normals.swap(index, last_index);
+            false
+        }
+    }
+
+    /// Same as `release`, but working on control FSM.
+    pub fn release_control(&mut self, control_box: &BasicMailbox<C>, channel_pos: usize) -> bool {
+        let s = self.control.take().unwrap();
+        control_box.release(s);
+        if control_box.len() == channel_pos {
+            true
+        } else {
+            match control_box.take_fsm() {
+                None => true,
+                Some(s) => {
+                    self.control = Some(s);
+                    false
+                }
+            }
+        }
+    }
+
+    /// Same as `remove`, but working on control FSM.
+    pub fn remove_control(&mut self, control_box: &BasicMailbox<C>) {
+        if control_box.is_empty() {
+            let s = self.control.take().unwrap();
+            control_box.release(s);
+        }
+    }
+}
+
+/// A handler that poll all FSM in ready.
+///
+/// A General process works like following:
+/// ```
+/// loop {
+///     begin
+///     if control is ready:
+///         handle_control
+///     foreach ready normal:
+///         handle_normal
+///     end
+/// }
+/// ```
+///
+/// Note that, every poll thread has its own handler, which doesn't have to be
+/// Sync.
+pub trait PollHandler<N, C> {
```

Comment:
Why not define `N` and `C` as associated types?

Reply:
It's short and neat.